### PR TITLE
feat(cds): 过期分支预览页按合并/放弃分流 + CI 镜像等待看门狗

### PR DESCRIPTION
**Pull request by inernoro**
## 摘要

PR 合并/关闭后分支被删，原先一律落「启动失败」错误页。本 PR 按分支墓碑（BranchTombstone.reason）分流成两种明确的中间页：
- `merged` → 「已合并到主分支」页 + 主按钮切换到主分支预览
- `abandoned` → 「分支已放弃」页 + 跳 PR

同时新增极速版「等待 CI 镜像」看门狗，解决分支卡在 waiting 状态永不部署的可观测性缺陷。

### CDS 后端（cds/src/）

- `index.ts`：新增 `buildBranchMergedPageHtml` / `buildBranchAbandonedPageHtml` 两个中间页生成函数，沿用 transit 页同款 shapeGridBg 动效网格背景 + 暖色双主题 token；新增 `startCiWaitWatchdog` 看门狗，监测 ciImageStatus='waiting' 超时（默认 15 分钟）并翻 failed + 写 ciImageError 归因；重构 `serveBranchGonePage` 先查分支墓碑分流，再落回泛化「启动失败」页；提取 `derivePreviewHost` 辅助函数供两处复用
- `services/state.ts`：新增 `recordRemovedBranch` 记录分支墓碑（previewSlug 为键），容量上限 200 条按 removedAt 淘汰最旧；新增 `getRemovedBranch` 查询；扩展 `updateBranchGithubMeta` 支持 ciWaitingSince / ciImageError 字段
- `services/github-webhook-dispatcher.ts`：PR closed 事件返 `tombstoneRequest`（reason='merged'/'abandoned'），delete 事件也返 tombstoneRequest（reason='abandoned'）；真 CI 完成事件到达时清掉 ciWaitingSince + ciImageError（允许 failed → ready 恢复）
- `routes/github-webhook.ts`：路由层据 tombstoneRequest 计算 previewSlug 并调 recordRemovedBranch 落库
- `types.ts`：新增 BranchTombstone / BranchRemovalReason 类型；扩展 BranchEntry 新增 ciWaitingSince / ciImageError 字段；扩展 CdsState 新增 removedBranches 字段

### CDS 前端（cds/web/src/）

- `pages/BranchListPage.tsx`：拆出 isCiWaiting / isCiFailed / isNeverDeployed 三态，修正 shouldShowStopReason 对齐后端 isStoppedBranch 口径（真有停止信号才显示停止面板）；新增「等待 CI 镜像」/ 「CI 镜像未就绪」两个状态块，展示 ciImageError 人类可读文案 + 构建链接

### 测试（cds/tests/）

- `services/state.test.ts`：新增 removed-branch tombstones 测试套件（record/get/persist/cap/merged 粘性）、CI 镜像字段 round-trip 测试
- `services/github-webhook-dispatcher.test.ts`：新增 PR merged / PR abandoned / raw branch delete 的 tombstoneRequest

https://claude.ai/code/session_01QiEHem1B2eURg786kbqsJc

---

> [!NOTE]
> **Medium Risk**
> Touches preview proxy routing, persisted state (`removedBranches`), and GitHub webhook semantics; regressions could misroute users or mislabel branch/CI status, but behavior is heavily tested and mostly user-facing HTML/SSE.
> 
> **Overview**
> **Expired preview UX** no longer defaults to a generic failure page. Webhooks persist **`BranchTombstone`** records (`merged` / `abandoned`) with lookup by preview slug, branch id, and aliases; **`serveBranchGonePage`** and **`ProxyService`** (via **`onBranchGone`**) serve new themed HTML for merged vs abandoned, including when a stopped entry still exists after PR close.
> 
> **Express CI observability**: a periodic **`startCiWaitWatchdog`** marks long **`ciImageStatus='waiting'`** branches as failed with **`ciImageError`**, events, and SSE-friendly field clears (`''` not `undefined`). Webhook/CI paths set **`ciWaitingSince`** / errors consistently. **`BranchListPage`** shows **等待 CI 镜像**, **CI 镜像未就绪**, and **待部署** instead of mislabeled stop rows.
> 
> Tests and debt/changelog docs cover tombstone edge cases (policy off, webhook ordering, slug reuse) and proxy tombstone diversion.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 46db2bd7b41e0133530cfd878e696ecdc791a615. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: cds/src/services/state.ts [hunk 1]
```diff
@@ -3101,6 +3106,93 @@ export class StateService {
...
+    // 复用兜底，合并后的自动删除是秒级，复用删除是小时级以上）。跨生命周期时既不保留 merged
+    // 粘性、也不承袭旧 PR 元数据/身份（否则旧 merged 页 + 旧 PR 链接会盖住当前 abandoned 状态，Codex P2）。
+    const existingAt = Date.parse(existing?.removedAt || '') || 0;
+    const incomingAt = Date.parse(tombstone.removedAt || '') || 0;
+    const differentPr = tombstone.prNumber != null && existing?.prNumber != null
+      && tombstone.prNumber !== existing.prNumber;
+    const farLater = existingAt > 0 && incomingAt > existingAt + 30 * 60 * 1000;
+    const crossLifecycle = !!existing && (differentPr || farLater);
+    if (existing?.reason === 'merged' && tombstone.reason === 'abandoned' && !crossLifecycle) {
+      return; // merged 粘性：同一生命周期的合并后 raw delete 不降级（跨生命周期复用则放行覆盖）
```
[the remark this answers is not in the document]
Comment (inernoro) [reply cds/src/services/state.ts:3137]: Accepted tradeoff — won't-fix (documented in `doc/debt.cds-removed-branch-pages.md` #7).

A raw branch `delete` webhook carries no `prNumber`, so within the post-merge window it is genuinely indistinguishable from a fast slug-reuse delete — there is no lifecycle identifier to key on. Any time window trades one boundary for the other:

- shorter → re-breaks the merge→auto-delete stickiness an earlier review round explicitly required (the common path: every merged PR fires `closed(merged)` then a `delete` seconds later);
- longer → the rare "reuse + delete within ~30 min" case flagged here.

The 30-min window deliberately favors the overwhelmingly-common merge-auto-delete. The deterministic half (a different `prNumber` from a later PR close) is already handled and unaffected. A fully deterministic fix would require threading the deleted branch's `createdAt` into the delete-path tombstone for lifecycle comparison — larger surface for marginal benefit; tracked as debt.

---
_Generated by [Claude Code](https://claude.ai/code)_
